File: tools/update_publications.py

from arb_update_resources import parse_args, get_irods_environment, setup_session
from irods.exception import NetworkException
from irods.models import Collection, CollectionMeta

def update_publications(session, updateDatacite="Yes", updateLandingpage="Yes", updateMOAI="Yes", package='*'):
    print(f"[UPDATE PUBLICATIONS] Start for {package}")
    with session:
        query = (
            session.query(Collection.name, CollectionMeta.name, CollectionMeta.value)  # Equivalent to selecting COLL_NAME
            .filter(CollectionMeta.name == 'org_' + 'vault_status')  # Equivalent to META_COLL_ATTR_NAME
            .filter(CollectionMeta.value == 'PUBLISHED')  # Equivalent to META_COLL_ATTR_VALUE
        )
        results = query.all()
        filtered_results = [result for result in results if result[Collection.name].startswith('/home/vault-')]
        for result in filtered_results:
            coll_name = result[0]
            
            if (package == '*' and re.match(r'/[^/]+/home/vault-.*', coll_name)) or (package != '*' and re.match(r'/[^/]+/home/vault-.*', coll_name) and coll_name == package):
                packages_found = True
                status, status_info = '', ''
                rule_update_publication(coll_name, update_datacite, update_landingpage, update_moai, status, status_info)
                print(f"{coll_name}: {status} {status_info}")
                
    if not packages_found:
        print(f"[UPDATE PUBLICATIONS] No packages found for {package}")
    else:
        print(f"[UPDATE PUBLICATIONS] Finished for {package}")

def main():
    args = parse_args()
    env = get_irods_environment()

    try:
        session = setup_session(env)
        update_publications(session)
    except NetworkException:
        print("Could not connect to iRODS sever ...")


if __name__ == '__main__':
    main()

File: tools/update_publications.r
import genquery
import re

def main(rule_args, callback, rei):
    callback.writeLine("stdout", "[UPDATE PUBLICATIONS] Start for {}".format(global_vars))
    update_datacite = global_vars["*updateDatacite"]
    update_landingpage = global_vars["*updateLandingpage"]
    update_moai = global_vars["*updateMOAI"]
    package = global_vars["*package"].strip('"')

    callback.writeLine("stdout", "[UPDATE PUBLICATIONS] Start for {}".format(package))

    coll_names = genquery.row_iterator(
        "COLL_NAME",
        "COLL_NAME like '%%/home/vault-%%' "
        "AND META_COLL_ATTR_NAME = 'org_vault_status' "
        "AND META_COLL_ATTR_VALUE = 'PUBLISHED'",
        genquery.AS_TUPLE,
        callback
    )

    packages_found = False
    for coll_name in coll_names:
        if ((package == '*' and re.match(r'/[^/]+/home/vault-.*', coll_name)) or (package != '*' and re.match(r'/[^/]+/home/vault-.*', coll_name) and coll_name == package)):
            packages_found = True
            status, status_info = '', ''
            res = callback.rule_update_publication(coll_name, update_datacite, update_landingpage, update_moai, status, status_info)
            callback.writeLine("stdout", "{}: {}".format(coll_name, res['arguments'][-2]))
            
    if not packages_found:
        callback.writeLine("stdout", "[UPDATE PUBLICATIONS] No packages found for {}".format(package))
    else:
        callback.writeLine("stdout", "[UPDATE PUBLICATIONS] Finished for {}".format(package))

input *updateDatacite="Yes", *updateLandingpage="Yes", *updateMOAI="Yes", *package="*"
output ruleExecOut
